File: instrumentation/src/test/java/io/opentelemetry/android/instrumentation/activity/ActivityCallbackTestHarness.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.activity;

import static org.mockito.Mockito.mock;

import android.app.Activity;
import android.os.Bundle;

class ActivityCallbackTestHarness {

    private final ActivityCallbacks callbacks;

    ActivityCallbackTestHarness(ActivityCallbacks callbacks) {
        this.callbacks = callbacks;
    }

    void runAppStartupLifecycle(Activity mainActivity) {
        // app startup lifecycle is the same as a normal activity lifecycle
        runActivityCreationLifecycle(mainActivity);
    }

    void runActivityCreationLifecycle(Activity activity) {
        Bundle bundle = mock(Bundle.class);

        callbacks.onActivityPreCreated(activity, bundle);
        callbacks.onActivityCreated(activity, bundle);
        callbacks.onActivityPostCreated(activity, bundle);

        runActivityStartedLifecycle(activity);
        runActivityResumedLifecycle(activity);
    }

    void runActivityStartedLifecycle(Activity activity) {
        callbacks.onActivityPreStarted(activity);
        callbacks.onActivityStarted(activity);
        callbacks.onActivityPostStarted(activity);
    }

    void runActivityPausedLifecycle(Activity activity) {
        callbacks.onActivityPrePaused(activity);
        callbacks.onActivityPaused(activity);
        callbacks.onActivityPostPaused(activity);
    }

    void runActivityResumedLifecycle(Activity activity) {
        callbacks.onActivityPreResumed(activity);
        callbacks.onActivityResumed(activity);
        callbacks.onActivityPostResumed(activity);
    }

    void runActivityStoppedFromRunningLifecycle(Activity activity) {
        runActivityPausedLifecycle(activity);
        runActivityStoppedFromPausedLifecycle(activity);
    }

    void runActivityStoppedFromPausedLifecycle(Activity activity) {
        callbacks.onActivityPreStopped(activity);
        callbacks.onActivityStopped(activity);
        callbacks.onActivityPostStopped(activity);
    }

    void runActivityDestroyedFromStoppedLifecycle(Activity activity) {
        callbacks.onActivityPreDestroyed(activity);
        callbacks.onActivityDestroyed(activity);
        callbacks.onActivityPostDestroyed(activity);
    }

    void runActivityDestroyedFromPausedLifecycle(Activity activity) {
        runActivityStoppedFromPausedLifecycle(activity);
        runActivityDestroyedFromStoppedLifecycle(activity);
    }

    void runActivityDestroyedFromRunningLifecycle(Activity activity) {
        runActivityStoppedFromRunningLifecycle(activity);
        runActivityDestroyedFromStoppedLifecycle(activity);
    }

    void runActivityRestartedLifecycle(Activity activity) {
        runActivityStartedLifecycle(activity);
        runActivityResumedLifecycle(activity);
    }
}

File: instrumentation/activity/src/test/java/io/opentelemetry/android/instrumentation/activity/ActivityCallbackTestHarness.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.activity

import android.app.Activity
import android.os.Bundle
import io.mockk.mockk

internal class ActivityCallbackTestHarness(
    private val callbacks: ActivityCallbacks,
) {
    fun runAppStartupLifecycle(mainActivity: Activity) {
        // app startup lifecycle is the same as a normal activity lifecycle
        runActivityCreationLifecycle(mainActivity)
    }

    fun runActivityCreationLifecycle(activity: Activity) {
        val bundle = mockk<Bundle>()
        callbacks.onActivityPreCreated(activity, bundle)
        callbacks.onActivityCreated(activity, bundle)
        callbacks.onActivityPostCreated(activity, bundle)
        runActivityStartedLifecycle(activity)
        runActivityResumedLifecycle(activity)
    }

    fun runActivityStartedLifecycle(activity: Activity) {
        callbacks.onActivityPreStarted(activity)
        callbacks.onActivityStarted(activity)
        callbacks.onActivityPostStarted(activity)
    }

    fun runActivityPausedLifecycle(activity: Activity) {
        callbacks.onActivityPrePaused(activity)
        callbacks.onActivityPaused(activity)
        callbacks.onActivityPostPaused(activity)
    }

    fun runActivityResumedLifecycle(activity: Activity) {
        callbacks.onActivityPreResumed(activity)
        callbacks.onActivityResumed(activity)
        callbacks.onActivityPostResumed(activity)
    }

    fun runActivityStoppedFromRunningLifecycle(activity: Activity) {
        runActivityPausedLifecycle(activity)
        runActivityStoppedFromPausedLifecycle(activity)
    }

    fun runActivityStoppedFromPausedLifecycle(activity: Activity) {
        callbacks.onActivityPreStopped(activity)
        callbacks.onActivityStopped(activity)
        callbacks.onActivityPostStopped(activity)
    }

    fun runActivityDestroyedFromStoppedLifecycle(activity: Activity) {
        callbacks.onActivityPreDestroyed(activity)
        callbacks.onActivityDestroyed(activity)
        callbacks.onActivityPostDestroyed(activity)
    }

    fun runActivityDestroyedFromPausedLifecycle(activity: Activity) {
        runActivityStoppedFromPausedLifecycle(activity)
        runActivityDestroyedFromStoppedLifecycle(activity)
    }

    fun runActivityDestroyedFromRunningLifecycle(activity: Activity) {
        runActivityStoppedFromRunningLifecycle(activity)
        runActivityDestroyedFromStoppedLifecycle(activity)
    }

    fun runActivityRestartedLifecycle(activity: Activity) {
        runActivityStartedLifecycle(activity)
        runActivityResumedLifecycle(activity)
    }
}
